### PR TITLE
ENH: Add aliases for commonly used dtype-like objects 

### PR DESCRIPTION
**Pull request by BvB93**
This PR adds (invariant) aliases for commonly used dtype-like objects, similar to what https://github.com/numpy/numpy/pull/18050 has done for array-likes.

Note that the aliases introduced herein are currently unused (beyond their introduction). 
This will change once we start adding dtype-support to the ufuncs, something that's on 
the to-do list for after the ufunc-based `ndarray` magic methods have been wrapped up (xref https://github.com/numpy/numpy/pull/18128 and https://github.com/numpy/numpy/pull/18228).

Secondly, it performs a small follow-up on https://github.com/numpy/numpy/pull/13578, _i.e._ objects with the `.dtype` attribute are 
now only considered dtype-like if aforementioned attribute returns a `np.dtype` instance.

Examples
----------
Example usage of the new `_DTypeLike<x>` aliases in annotating a simplified version of `np.bitwise_or`.
``` python
from __future__ import annotations
from typing import Any, overload

import numpy as np
import numpy.typing as npt

@overload
def bitwise_or(
    __x1: npt._ArrayLikeBool_co, 
    __x2: npt._ArrayLikeBool_co,
    *, dtype: npt._DTypeLikeBool | None = ...,
) -> np.ndarray[Any, np.dtype[np.bool_]]: ...
@overload
def bitwise_or(
    __x1: npt._ArrayLikeUInt_co, 
    __x2: npt._ArrayLikeUInt_co, 
    *, dtype: npt._DTypeLikeUInt | None = ...,
) -> np.ndarray[Any, np.dtype[np.unsignedinteger[Any]]: ...
@overload
def bitwise_or(
    __x1: npt._ArrayLikeInt_co, 
    __x2: npt._ArrayLikeInt_co, 
    *, dtype: npt._DTypeLikeInt | None = ...,
) -> np.ndarray[Any, np.dtype[np.signedinteger[Any]]: ...
```

### CODE REVIEW

File: numpy/typing/_dtype_like.py [hunk 1]
```diff
@@ -79,3 +124,107 @@ class _SupportsDType(Protocol):
...
+    Type[np.floating],
+    "np.dtype[np.floating]",
+    "_SupportsDType[np.dtype[np.floating]]",
+    _Float16Codes,
+    _Float32Codes,
+    _Float64Codes,
+    _HalfCodes,
+    _SingleCodes,
+    _DoubleCodes,
+    _LongDoubleCodes,
```
Comment (BvB93) [inline numpy/typing/_dtype_like.py:179]: A typical example of the newly introduced `_DTypeLike<X>` unions:
* A `np.generic` subclass.
* A builtin counterpart to aforementioned `np.generic` subclass.
* A `np.dtype` instance.
* An object with the `dtype` attribute.
* Various string literals with relevant character codes (`f`, `float`, `f4`, _etc._).

File: numpy/typing/_dtype_like.py [hunk 2]
```diff
@@ -79,3 +124,107 @@ class _SupportsDType(Protocol):
...
+    _UIntPCodes,
+    _UIntCodes,
+    _ULongLongCodes,
+]
+_DTypeLikeInt = Union[
+    Type[int],
+    Type[np.signedinteger],
+    "np.dtype[np.signedinteger]",
+    "_SupportsDType[np.dtype[np.signedinteger]]",
+    _UInt8Codes,
```
Comment (rgommers) [inline numpy/typing/_dtype_like.py:157]: typo here `Uint` -> `Int

**Review by rgommers — APPROVED** [review]
Thanks @BvB93, this looks good to me modulo the one typo.

File: numpy/typing/_dtype_like.py (hunk 2 above)
Comment (BvB93) [reply numpy/typing/_dtype_like.py]: Good catch! Fixed as of https://github.com/numpy/numpy/pull/18236/commits/d788788f8233d96e647c64bd4070305146b9ff92.

**Comment by rgommers** [comment]
Merged, thanks!

**Pull request merged**